src/main/java/com/github/packageurl/PackageURL.java

```diff
@@ -347,22 +344,31 @@ private String validatePath(final String value, final boolean isSubpath) throws
         return validatePath(value.split("/"), isSubpath);
     }
 
-    private String validatePath(final String[] segments, final boolean isSubpath) throws MalformedPackageURLException {
+    private static boolean isValidSegment(String segment, final boolean isSubpath) {
+        return (!isSubpath || (!segment.isEmpty() && !".".equals(segment) && !"..".equals(segment)));
+    }
+
+    private static String validatePath(final String[] segments, final boolean isSubpath) throws MalformedPackageURLException {
         if (segments == null || segments.length == 0) {
             return null;
         }
+
         try {
             return Arrays.stream(segments)
                     .map(segment -> {
-                        if (isSubpath && ("..".equals(segment) || ".".equals(segment))) {
-                            throw new ValidationException("Segments in the subpath may not be a period ('.') or repeated period ('..')");
-                        } else if (segment.contains("/")) {
-                            throw new ValidationException("Segments in the namespace and subpath may not contain a forward slash ('/')");
-                        } else if (segment.isEmpty()) {
-                            throw new ValidationException("Segments in the namespace and subpath may not be empty");
+                        if (!isSubpath) {
+                            if ("..".equals(segment) || ".".equals(segment)) {
+                                throw new ValidationException("Segments in the namespace may not be a period ('.') or repeated period ('..')");
+                            } else if (segment.contains("/")) {
+                                throw new ValidationException("Segments in the namespace and subpath may not contain a forward slash ('/')");
+                            } else if (segment.isEmpty()) {
+                                throw new ValidationException("Segments in the namespace and subpath may not be empty");
+                            }
                         }
                         return segment;
-                    }).collect(Collectors.joining("/"));
+                    })
+                    .filter(segment1 -> isValidSegment(segment1, isSubpath))
```

Comment:
```suggestion
```

Discardable segments were already discarded in `parsePath(String, boolean)` or an exception was thrown in the lines above.

Reply:
Again, code was changed per the subject line of this pull request based on the spec:

> Discard any empty string segment from that split
> Discard empty, '.' and '..' segments

It will only throw an exception right now for `namespace`, but not `subpath`.

Reply:
I probably should have selected more context for this code snippet. This is part of `validatePath`, which is called after `parsePath`. So:

1. `parsePath` removes discardable (invalid) segments.
2. `validatePath` throws on the invalid segments that remain (i.e. those with `/`).
3.  on line 370, we already know that all segments are valid.

Reply:
Can you just clarify what you are saying. Do you think the current code is 100% correct, or not?

Again, please try the current code with that test suite. I do not think they pass with the current code, which is the whole point of opening this pull request.

https://github.com/package-url/purl-spec/blob/7f7e82f73c38a4a339f88abf1f2aa031e9c7af23/test-suite-data.json#L50-L73

Reply:
> Can you just clarify what you are saying. Do you think the current code is 100% correct, or not?

I think line 370 is not useful, since all the invalid or discardable segments have already been removed by the callers of `validatePath`.

Reply:
Has this been fixed in the latest version. No changes have been made to where it filters.